Use standard linting tools across all of our projects (#153)
dxw/tech-team-rfcs: rfc-035-use-standard-linting-tools-across-all-our-projects.md

# Use standard linting tools across all of our projects

## Summary

> Where available, all new dxw-owned or maintained repositories SHOULD use a

Reviewer: Have some words been snipped from here? This sentence doesn't parse for me.

Author: Whoops - good catch!

# Use standard linting tools across all of our projects

## Summary

> Where available, all new dxw-owned or maintained repositories SHOULD use a

Reviewer 2: At the moment this just says that _a_ linter should be used, rather than the specific linter we've picked for that language, so I think we can make this language a bit stronger.

Reviewer 2: We should probably also add an option to break this rule with an ADR for situations where there is a good reason not to follow it (eg. the client has a linter they prefer)

Reviewer 3: I think use of "SHOULD" allows for breaking the rule if needed.

Reviewer 3: Also, I think this RFC is "use any linter". I'd welcome RFCs for specific linter choices, but I don't think it needs to be part of this RFC.

> linter for the programming language the repo is written in, where one is

Reviewer 4: If you're in Rails, you want to be linting Ruby and JS, so I've tweaked the wording a little.

```suggestion
linter for the major programming languages the repo is written in, where
```

available.

## Problem

As developers, we commonly have different styles of writing code. Unchecked this
can lead to inconsistencies in style, which over time can make code messy, and
difficult to maintain and review. It can also lead to bikeshedding over code
style in code reviews, which can cause conflict and slow down code review.

## Proposal

Where available, all new dxw-owned or maintained repositories SHOULD use a
standard linting tool to make sure code meets a standard style. Examples of this

> include (but are not limited to):

Reviewer 3: I suggest we further restrict to "an opinionated code linter and formatter". One of the motivations of this is to reduce the number of conversations about what the rules should be, and opinionated tools help by making those decisions for us. This is the main reason why I've pushed for Prettier for JS. If you can get over conforming to someone else's style, you stop having to think or talk about style ever again.

> - [Standard.rb](https://github.com/testdouble/standard)

Reviewer 2: Is there a reason we have chosen these specific linters over other options?

Reviewer 5: Black (which is an excellent choice) feels like it cleaves the jobs of a linter neatly into two: formatting the code (which it does excellently) and code style advice and enforcement (which it deftly avoids). There may be a role for PyLint or similar.

Reviewer 4: @jdudley1123 I think these are just examples, rather than explicit choices.

> - [ESLint](https://eslint.org/)

Reviewer 4: I's suggest changing this to the Standard equivalents where available. I know it's just an example, but more standard is a good thing.
```suggestion
- [Standard for Javascript](https://github.com/standard/standard)
- [Standard for Typescript](https://github.com/standard/ts-standard)
```

- [PHP-CS-Fixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer)
- [Black](https://github.com/psf/black)

When a linter is installed, each package MUST run a check using the linting tool

> as part of the default CI build (preferably before the main test suite runs). We

Reviewer 4: I disagree here. I think the linter should run *after* the test suite. It's incredibly frustrating to be looking for test failures and have something fail on the linter, which you then have to fix before finding out that your code is wrong in more fundamental ways. I prefer to run the linter after the tests have passed, including on CI.

MUST ensure a CI build fails if the linting fails.

We MUST add instructions to a project's README on how to run the tool in
development, before committing code to the repository. This MAY also include
instruction on how to run any autofixing tasks, if any are available.

## Next steps

We don't need to take any further action at this time. Standard.rb is already
added to the Rails template by default, and ESlint is already in the React

> template, so this is just codifying what we are already doing.

Reviewer 4: If we want to use StandardTS, we will need to change the template.

> template, so this is just codifying what we are already doing.

Reviewer 3: React template is archived and dead.

# Use standard linting tools across all of our projects

## Summary

Where available, all new dxw-owned or maintained repositories SHOULD use a
linter for the major programming languages the repo is written in.

## Problem

As developers, we commonly have different styles of writing code. Unchecked this
can lead to inconsistencies in style, which over time can make code messy, and
difficult to maintain and review. It can also lead to bikeshedding over code
style in code reviews, which can cause conflict and slow down code review.

## Proposal

Where available, all new dxw-owned or maintained repositories SHOULD use an
opinionated code linter and formatter to make sure code meets a standard style.
Examples of this include (but are not limited to):

- [Standard.rb](https://github.com/testdouble/standard)
- [Standard for Javascript](https://github.com/standard/standard)
- [Standard for Typescript](https://github.com/standard/ts-standard)
- [PHP-CS-Fixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer)

> - [Black](https://github.com/psf/black)

Reviewer 5: Super minor thing: I think it's worth appending `(for Python)` to the end of the Black line -- every other one is semi-obvious which language it's for.

Author: Good call 👍

# Use standard linting tools across all of our projects

## Summary

Where available, all new dxw-owned or maintained repositories SHOULD use a

> linter for the major programming languages the repo is written in, where one is
> available.

Reviewer 3: ```suggestion
linter for the major programming languages the repo is written in.
```

Author: 👍
